src/treebuilder.rs

```diff
@@ -0,0 +1,207 @@
+use libc::{c_int, c_void};
+use std::marker;
+
+use {panic, raw, tree, Error, Oid, Repository, TreeEntry};
+use util::{Binding, IntoCString};
+
+/// Constructor for in-memory trees
+pub struct TreeBuilder<'repo> {
+    raw: *mut raw::git_treebuilder,
+    _marker: marker::PhantomData<&'repo Repository>,
+}
+
+impl<'repo> TreeBuilder<'repo> {
+    /// Clear all the entries in the builder
+    pub fn clear(&mut self) {
+        unsafe { raw::git_treebuilder_clear(self.raw) }
+    }
+
+    /// Get the number of entries
+    pub fn len(&self) -> usize {
+        unsafe { raw::git_treebuilder_entrycount(self.raw) as usize }
+    }
+
+    /// Get en entry from the builder from its filename
+    pub fn get<P: IntoCString>(&self, filename: P) -> Result<Option<TreeEntry>, Error> {
+        let filename = try!(filename.into_c_string());
+        unsafe {
+            let ret = raw::git_treebuilder_get(self.raw, filename.as_ptr());
+            if ret.is_null() {
+                Ok(None)
+            } else {
+                Ok(Some(tree::entry_from_raw_const(ret)))
+            }
+        }
+    }
+
+    /// Add or update an entry in the builder
+    ///
+    /// No attempt is made to ensure that the provided Oid points to
+    /// an object of a reasonable type (or any object at all)
+    pub fn insert<P: IntoCString>(&mut self, filename: P, oid: Oid,
+                                  filemode: i32) -> Result<TreeEntry, Error> {
+        use raw::FromPrimitive;
+        let filename = try!(filename.into_c_string());
+        let filemode = raw::git_filemode_t::from_i32(filemode)
+            .expect("Invalid filemode passed to TreeBuilder::insert");
+
+        let mut ret = 0 as *const raw::git_tree_entry;
+        unsafe {
+            try_call!(raw::git_treebuilder_insert(&mut ret, self.raw, filename,
+                                                  oid.raw(), filemode));
+            Ok(tree::entry_from_raw_const(ret))
+        }
+    }
+
+    /// Remove an entry from the builder by its filename
+    pub fn remove<P: IntoCString>(&mut self, filename: P) -> Result<(), Error> {
+        let filename = try!(filename.into_c_string());
+        unsafe {
+            try_call!(raw::git_treebuilder_remove(self.raw, filename));
+        }
+        Ok(())
+    }
+
+    /// Selectively remove entries from the tree
+    pub fn filter<F>(&mut self, mut filter: F)
+                     where F: FnMut(&TreeEntry) -> FilterResult {
```

Comment:
I'd be fine having this just return a `bool` as `filter`'s pretty common with iterators already so the enum may not be buying mcuch